Enh/surface plotting2

juhuntenburg opened the pull request:
Adding surface visualization along with example and data fetcher for NKI enhanced surface data.

AlexandreAbraham commented:
Juste after I left, you Hotta ne kidding me :D

GaelVaroquaux commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,227 @@
+            Parameters
+            ----------
+            surf_mesh: Surface object (to be defined)
+            hemi: Hemisphere to display
```
I'd prefer if the acceptable value for "hemi" would be "left" and "right", rather than "lh" and "rh", which are a bit cryptic.

juhuntenburg replied:
Changed

AlexandreAbraham commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+# Retrieve the data
+nki_dataset = datasets.fetch_nki_enhanced_surface(n_subjects=1)
+
+fsaverage5_pial = nki_dataset['fsaverage_pial'][0][0]
```
I guess that the final `[0]` means `lh`. I think we should rather make two fields `fsaverage_pial_left` and `fsaverage_pial_right`.

KamalakerDadi replied:
yes that would be more clear.

juhuntenburg replied:
Agree, I changed it.

GaelVaroquaux replied:
> I guess that the final [0] means lh. I think we should rather make two fields
> fsaverage_pial_left and fsaverage_pial_right.

+1!

AlexandreAbraham commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,227 @@
+                surf_mesh.endswith('inflated')):
+            coords, faces = nibabel.freesurfer.io.read_geometry(surf_mesh)
+        elif surf_mesh.endswith('gii'):
+            coords, faces = gifti.read(surf_mesh).getArraysFromIntent(1008)[0].data, \
```
You can use `nibabel.nifti1.intent_codes['NIFTI_INTENT_POINTSET']` rather than `1008`

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+"""
+Demonstrate basic surface plotting with plot_surf_stat_map
+==========================================================
```
May be you can write few lines about what this example does. For example, I don't know what ```nki``` is ?
stating that we are fetching so and so datasets and visualizing them.

juhuntenburg replied:
Yes, I will add a description both of the dataset and the example asap.

GaelVaroquaux replied:
> May be you can write few lines about what this example does. For example, I
> don't know what nki is ?
> stating that we are fetching so and so datasets and visualizing them.

And tell us that this is resting activation.

I just realized something: is it possible to download the FS standard
parcellations from the web? That would make an awesome visualization
example.

juhuntenburg replied:
The parcellations are provided with the FS distribution, so I guess we could use them? The problem is that this function is not exactly optimized to plot labels / rois: the boundaries will be blurred because of the interpolation of vertex values to face values. An easy way around this should be to use the median value instead of the mean. But I thought this would go into a separate function (something similar to plot_roi?) and thus separate example. 
I agree that the data that is used now not a great example, because it isn't a stat map either, but just some random volume of a time series. So maybe I should look into creating an example map from the nki dataset? Or would it be an option to just use some map from Neurovault for now?

GaelVaroquaux replied:
> The parcellations are provided with the FS distribution, so I guess we could
> use them?

Well, we cannot rely on people having FS installed, so we would need to
provide downloaders, and for this we would need to check that there
exists URLs where we can download them.

> median value instead of the mean. But I thought this would go into a separate
> function (something similar to plot_roi?) and thus separate example.

I agree. Let's do this later.

> So maybe I should look into creating an example map from the nki
> dataset?

Let's worry about this later.

juhuntenburg replied:
Or just upload them to the example dataset just as we did for the fsaverage mesh?

GaelVaroquaux replied:
> Or just upload them to the example dataset just as we did for the fsaverage
> mesh?

It's an option indeed.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+Demonstrate basic surface plotting with plot_surf_stat_map
+==========================================================
+
+See :ref:`plotting` for more details.
```
pin point reference may be :func:nilearn.plotting.plot_surf_stat_map ?

GaelVaroquaux replied:
> pin point reference may be :func:nilearn.plotting.plot_surf_stat_map ?

I think that this is OK, as long as there is a section about surface
visualization in the docs.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+from nilearn import plotting
+from nilearn import datasets
+
+# Retrieve the data
```
All examples start with ############### to have head line representation. If you want you can have a look at other examples.

juhuntenburg replied:
Done

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+plotting.plot_surf_stat_map(fsaverage5_pial, 'lh', stat_map=stat_map,
+                            output_file='/tmp/plot_surf_stat_map.png')
+
+plt.show()
```
you can use ```plotting.show()```

juhuntenburg replied:
Oh good, changed it.

GaelVaroquaux replied:
> you can use plotting.show()

Yes. We do this to avoid having to teach beginners to import matplotlib.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+
+See :ref:`plotting` for more details.
+"""
+import matplotlib.pyplot as plt
```
may be its not necessary to import matplotlib

juhuntenburg replied:
Yes, no it isn't necessary anymore.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,58 @@
+
+
+# display unthresholded stat map in lateral and medial view
+plotting.plot_surf_stat_map(fsaverage5_pial, 'lh', stat_map=stat_map)
```
adding ```title``` would be nice I guess ?

KamalakerDadi commented:
Example looks great. Just minor details on example ```plot_surf_stat_map```

KamalakerDadi commented on nilearn/datasets/__init__.py:
```diff
@@ -7,7 +7,8 @@
                    fetch_localizer_calculation_task, fetch_mixed_gambles,
-                   fetch_megatrawls_netmats, fetch_cobre)
+                   fetch_megatrawls_netmats, fetch_cobre,
+                   fetch_nki_enhanced_surface)
```
you may also need to add in ```doc/modules/reference.rst```

juhuntenburg replied:
Done

GaelVaroquaux commented:
@juhuntenburg : if I look at one of your images, you have a huge amount of not useful whitespace around a brain:

https://circle-artifacts.com/gh/nilearn/nilearn/1298/artifacts/0/home/ubuntu/nilearn/doc/_build/html/_images/sphx_glr_plot_surf_stat_map_004.png

We need to do something about this.

juhuntenburg commented:
Yes I agree about the white space. My colleague uses another function to crop the image and combine medial and lateral view into one display: https://github.com/soligschlager/brainsurfacescripts/blob/temp/plotting.py#L343-L420 Would some version of that be an option? Then I would look into it with her. 

Or I will check whether it is possible to avoid the white space in matplotlib in the first place

GaelVaroquaux commented:
> the image and combine medial and lateral view into one display: https://
> github.com/soligschlager/brainsurfacescripts/blob/temp/plotting.py#L343-L420
> Would some version of that be an option?

Hum, that solution isn't great: it saves to a raster image after
reloading.

> Or I will check whether it is possible to avoid the white space in matplotlib
> in the first place

That should be possible. In the volumetric plotting code we compute the
bounds of the objects that we plot, and we use them to adapt the bounds
of the plot.

GaelVaroquaux commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,227 @@
+                                          symmetric_cbar, kwargs)
+
+            if threshold is not None:
+                kept_indices = np.where(abs(stat_map_faces) >= threshold)[0]
```
Should the "abs" be an "np.abs"?

KamalakerDadi commented:
Hi @juhuntenburg Any progress on this ?

juhuntenburg commented:
Hi @KamalakerDadi: ah yes, good reminder.

1. White space around the figure:
I looked into this but didn't find a solution yet. In the volumetric plotting code, I figure this is the step to adapt the bounds: https://github.com/nilearn/nilearn/blob/master/nilearn/plotting/displays.py#L92-L94 ? 
But since I don't use imshow here I don't see how to translate it directly. I tried to change the bounding box of the 3d axis with ax.set_clip_box but it didn't help. Maybe one of you has an idea that they could point me to?

2. Example / documentation: 
I would like to replace the example which plots an arbitrary volume of a time series with one that plots seed-based connectivity of the example dataset. I will look into this asap and then also write a proper documentation of the example. 

Are there any other crucial points for now?

KamalakerDadi commented:
Thanks @juhuntenburg 

>White space around the figure:

I will try to search for some solutions/hints.

>Example / documentation: 

yes, please. It looks interesting.

juhuntenburg commented:
I have reworked the example. For this the Destrieux parcellation (shipped with Freesurfer) would need to be added to the Nilearn NKI dataset. For testing you can find the file here: https://www.dropbox.com/s/j66g4ljqaj5hz3a/lh.aparc.a2009s.annot?dl=0
Happy for feedback!

juhuntenburg commented:
I also want to make a second function plot_surf_roi (very similar just using the median instead of mean value for the face color and some other defaults) and display the seed region as well.

GaelVaroquaux commented:
> I have reworked the example. For this the Destrieux parcellation (shipped with
> Freesurfer) would need to be added to the Nilearn NKI dataset.

I am +1 adding the Destrieux parcellation, but in a different dataset,
and with a different downloader. Ideally, we should also be able to have
an example plotting it.

GaelVaroquaux commented:
> I also want to make a second function plot_surf_roi (very similar just using
> the median instead of mean value for the face color and some other defaults)

I am not sure that I understood what this function does, and its purpose.
Would you mind detailling a little more. Cheers.

juhuntenburg commented:
The current function is not ideal for plotting labels (atlas, roi) because it interpolates the values and doesn't create clean borders. This can easily be changed by using the median value instead of the mean to calculate the face colors. 
Does that make more sense?

GaelVaroquaux commented:
> The current function is not ideal for plotting labels (atlas, roi) because it
> interpolates the values and doesn't create clean borders. This can easily be
> changed by using the median value instead of the mean to calculate the face
> colors.
> Does that make more sense?

It makes more sense in terms of the purpose. However, the median value
surprised: if those are labels, than the mode should be used, rather than
the median (ie the label number that is most represented). Now, I realize
that if we are talking triangles, there are only three values, in which
case, either the mode is undefined, or it is equal to the median.

KamalakerDadi commented:
Can you please rebase and resolve the conflicts ?

I can review your PR again this week.

juhuntenburg commented:
Yes of course, just quick question to be on the save side: I updated my master and then rebased the surface plotting branch to master. Now when I want to push the rebased branch to its remote I would need to do push --force because it cannot fast-forward anymore. Is it ok to do that or is there a better way?

juhuntenburg commented:
@GaelVaroquaux: would you suggest adding the Destrieux dataset fetcher, an atlas plotting version of the plotting function and an example in this PR or can it wait for later? I could also avoid using the atlas in the current example and just hard code the seed labels.

bthirion commented:
Yes, in that case you push --force. 
Thanks for keeping this PR alive !

juhuntenburg commented:
Done, thanks @bthirion !

GaelVaroquaux commented:
> Now when I want to push the rebased branch to its remote I would need
> to do push --force because it cannot fast-forward anymore. Is it ok to
> do that or is there a better way?

It's OK, it's your own private branch

GaelVaroquaux commented:
> @GaelVaroquaux: would you suggest adding the Destrieux dataset fetcher, an
> atlas plotting version of the plotting function and an example in this PR or
> can it wait for later? I could also avoid using the atlas in the current
> example and just hard code the seed labels.

The problem of your PR is that it includes only one usease, hence it is
hard to review the API/usability of the code. Therefore, I believe that
it is useful that you add to this PR the new dataset + function +
example.

What we are doing here is sketching the API and usability of nilearn with
surface. We will need several examples and usecases for that.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,132 @@
+same hemisphere is calculated using Pearson product-moment correlation
+coefficient.
+
+The :func:`nilearn.plotting.surf_plotting.plot_surf_stat_map` function is used
```
May be you can use here: `nilearn.plotting.plot_surf_stat_map`. Thats the reason for circleCI failure I guess.

juhuntenburg replied:
Done!

juhuntenburg commented:
@GaelVaroquaux: Makes sense, I'll work on it as soon as I can!

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,132 @@
+
+# Destrieux parcellation left hemisphere in fsaverage5 space
+# TO DO: include in nki_dataset for fetching
+destrieux = '/home/julia/nilearn_testing/lh.aparc.a2009s.annot'
```
reason for CircleCI failure. 

If you are looking for a destrieux atlas, why don't you pick from nilearn ? ```fetch_atlas_destrieux_2009``` 
Is it not the same you are using here ?

KamalakerDadi replied:
Ok may be its not I have seen the reference after commenting. Destrieux et al, (2010)

juhuntenburg replied:
Actually I am not sure it is a different parcellation (It is called 2009 within Freesurfer but the final atlas was published 2010), but anyhow the main difference is that this is surface space and the one in Nilearn is a volume, right?

KamalakerDadi replied:
May be can you have a look at the maps. I could not differentiate between surface and volume. But this may help fetched from that atlas description.

Notes
-----
The 'Destrieux' cortical atlas is based on a parcellation scheme
that first divided the cortex into gyral and sulcal regions,
the limit between both being given by the curvature value of the surface.
A gyrus only includes the cortex visible on the pial view,
the hidden cortex (banks of sulci) are marked sulcus.
The result is a complete labeling of cortical sulci and gyri.

References
----------
Fischl, Bruce, et al. "Automatically parcellating the human cerebral cortex."
Cerebral cortex 14.1 (2004): 11-22.

Destrieux, C., et al. "A sulcal depth-based anatomical parcellation
of the cerebral cortex." NeuroImage 47 (2009): S151.

juhuntenburg replied:
So the one in nilearn is definitely a volume (you can load it in MNI space). What I use here is basically just a dictionary associating nodes indices in fsaverage5 space with labels of the Destrieux parcellation, this is what is saved in Freesurfer's standard .a2009s.annot output. I am pretty sure the underlying parcellation is the same, but afaik there is no trivial way to translate one into the other (i.e. without using additional tools like the ones from Freesurfer) so that I could simply use the one that is already in Nilearn on the surface.

GaelVaroquaux replied:
> What I use here is basically just a dictionary associating nodes indices in
> fsaverage5 space with labels of the Destrieux parcellation, this is what is
> saved in Freesurfer's standard .a2009s.annot output. I am pretty sure the
> underlying parcellation is the same, but afaik there is no trivial way to
> translate one into the other

Indeed

GaelVaroquaux commented on doc/modules/reference.rst:
```diff
@@ -82,6 +82,7 @@ uses.
    fetch_localizer_calculation_task
    fetch_miyawaki2008
    fetch_nyu_rest
+   fetch_nki_enhanced_surface
```
Should we call this function "fetch_surf_nki_enhance"? Just like we have prefixed all the atlas fetching functions with fetch_atlas?

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1640,3 +1640,176 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+                           archive % (nitrc_ids[i], ids[i], 'lh'),
+                           {'move': func % (ids[i], ids[i], 'left')}
+                            )],
+                          resume=resume, verbose=verbose)
```
This usage of the low_level "_fetch_files" function is telling me that we probably need an additional data fetcher, to answer this specific usecase.

GaelVaroquaux replied:
Sorry. Stupid. I thought that I was looking at an example. Please forget this comment.

GaelVaroquaux commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,237 @@
+
+            Parameters
+            ----------
+            surf_mesh: Surface object (to be defined)
```
This part of the docstring still needs to be finished.

juhuntenburg commented:
Hi, I've been working on the additional function and starting to write the data fetcher for the Destrieux surface atlas. I was wondering if I should include the fsaverage5 surface giftis in that fetcher as well (without them I can't plot the atlas), like I did in the nki fetcher. Or, if there should be an independent fetcher for fsaverage5 (and I remove these files from the nki fetcher as well)?

GaelVaroquaux commented:
I think that an independent fsaverage fetcher seems right.

juhuntenburg commented:
Ok cool, what is the best way to create the new datasets on nitrc? Can I get access to do that?

GaelVaroquaux commented:
> Ok cool, what is the best way to create the new datasets on nitrc? Can I get access to do that?

Ask @AlexandreAbraham, if he is in the same room than you. Else, ping him

juhuntenburg commented:
Ah Vienna brainhack, no not there, will contact him

AlexandreAbraham commented:
Creating a new dataset on nitrc requires some "click click" actions. I can do it for you.

juhuntenburg commented:
(Please ignore the recent changes for now, working on finishing this with @Joaoloula at brainhack and will then rebase etc.)

KamalakerDadi commented:
>(Please ignore the recent changes for now, working on finishing this with @Joaoloula at brainhack and will then rebase etc.)

Thank you for getting back to this. Feel free to ping us if it is ready for reviewing.

juhuntenburg commented:
Hi, so we addressed the following points:
* Splitting the surface plotting function in a main function (plot_surf) and two functions (plot_surf_stat_map and plot_surf_roi) which call plot_surf with appropriate defaults.
* Including a second example plotting the Destrieux parcellation on the fsaverage 5 surface
* Adding data fetchers for the Destrieux annot files as well as the fsaverage5 surface files
* Documentation

GaelVaroquaux commented:
Coverall reports that the test coverage has significantly decreased. Is that spurious, or actually the case? Iow, is the new code well tested?

GaelVaroquaux commented:
As the circleci failure shows, the docstring of fetch_surf_fsaverage5 is not well formatted.

Joaoloula commented:
On the coverage department, we can write tests for the new code if you think it's a good idea. As for the fetch_surf_fsaverage5 docstring formatting, I'm under the impression it's just an empty line missing before the Parameters, can you confirm this before I commit?

KamalakerDadi commented:
>On the coverage department, we can write tests for the new code if you think it's a good idea.

Its always a good idea to test the written code.

>As for the fetch_surf_fsaverage5 docstring formatting, I'm under the impression it's just an empty line missing before the Parameters, can you confirm this before I commit?

Can you check on your machine by cd doc folder in nilearn and run ```make html-strict``` ? You will come to know whether failure is fixed or not.

juhuntenburg commented:
There are no tests whatsoever for this code, yet. To me that seemed a bit premature before I'd be more confident the functions would stay roughly as they are. Happy to start writing tests after this round of review, as I guess the remaining issues won't change the existing code dramatically.

juhuntenburg commented:
One thing I'd still like to discuss is if there will be something like a Nilearn mesh object (a niimg equivalent). Since there are so many surface file formats, to me it makes sense to allow users to figure out how to load their personal favourite into Python (the ones not supported by Nibabel that is), create a mesh object and then input that to the Nilearn functions directly.
I usually use a simple dictionary containing an entry for vertex coordinates and one for triangle indices, but happy to discuss alternatives.

bthirion commented:
There has been not previous attempt to work with meshes in nilearn, so there is no strategy on that topic so far. For the moment I would keep things as light as possible. I think that a mesh is nothing more than a set of coordinates + a set of triplets representing triangles.

juhuntenburg commented:
Precisely. However, the advantage of keeping these two sets in a single dictionary would be that there is only one object to pass between functions, rather than two, but maybe not such a big deal.

GaelVaroquaux commented:
In terms of design, I feel that it is always good to start will little
abstractions, and add some as we go. So, for now, I think that I would
prefer too arguments. As there is with a stats map and it's background
image.

juhuntenburg commented:
Sure I see the point. Just to be clear, it would then be three to four arguments:  the vertex coordinates, the face triplets, the stat map, and optionally the background (sulcal depth) map

juhuntenburg commented:
Sure I see the point. Just to be clear, it would then be three to four arguments:  the vertex coordinates, the face triplets, the stat map, and optionally the background (sulcal depth) map

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+
+def fetch_atlas_surf_destrieux(data_dir=None, url=None,
+                               resume=True, verbose=1):
+    """Download and load the Destrieux cortical atlas (dated 2009)
```
I would write "Download and load Destrieux et al, 2010 cortical atlas." and after one line empty space

I would write "This atlas returns 74 sulco-gyral structures per hemisphere distributed with Freesurfer in fsaverage5 surface space."

I choose to update with 2010 paper. The annotations are described well in this paper. At some point, users might ask about the reference point from where we have picked those annotations.

juhuntenburg replied:
Done

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    Parameters
+    ----------
+    data_dir: string, optional
+        Path of the data directory. Use to forec data storage in a non-
```
type forec -> force

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    ----------
+    data_dir: string, optional
+        Path of the data directory. Use to forec data storage in a non-
+        standard location. Default: None (meaning: default)
```
May be we can remove one default

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    -------
+    data: sklearn.datasets.base.Bunch
+        dictionary-like object, contains:
+        - Cortical ROIs, lateralized or not (maps)
```
Returns section needs to be updated. What we return here are path to annotations in .annot format.
I would update with the same as what you have written in destrieux_surface.rst file.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    data: sklearn.datasets.base.Bunch
+        dictionary-like object, contains:
+        - Cortical ROIs, lateralized or not (maps)
+        - Labels of the ROIs (labels)
```
This can be removed I guess.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    url: string, optional
+        Download URL of the dataset. Overwrite the default URL.
+
+    Returns
```
Why don't we return by loading them rather than path to .annot file ?

juhuntenburg replied:
We would need to import nibabel to load the annot file. Should we do this?

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    Fischl, Bruce, et al. "Automatically parcellating the human cerebral
+    cortex." Cerebral cortex 14.1 (2004): 11-22.
+
+    Destrieux, C., et al. "A sulcal depth-based anatomical parcellation
```
I would update with 2010 reference paper.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+        url = "https://www.nitrc.org/frs/download.php/"
+
+    dataset_name = 'destrieux_surface'
+    opts = dict()
```
opts is not use anywhere ?

KamalakerDadi commented on nilearn/datasets/description/destrieux_surface.rst:
```diff
@@ -0,0 +1,29 @@
+
+Notes
+-----
+Anatomical parcellation of the cortex based on sulco-gyral patterens 
```
typo patterens -> patterns

KamalakerDadi commented on nilearn/datasets/description/destrieux_surface.rst:
```diff
@@ -0,0 +1,29 @@
+References
+----------
+
+Destrieux, C., et al. "A sulcal depth-based anatomical parcellation
```
On the top you have cited 2010 paper. But, here you referred 2009 paper.

KamalakerDadi commented on nilearn/datasets/description/destrieux_surface.rst:
```diff
@@ -0,0 +1,29 @@
+Content
+-------
+    :'annot_left': parcellation on left hemisphere in freesurfer annot format
+    :'annot_right': parcellation on left hemisphere in freesurfer annot format
```
left -> right

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+
+def fetch_surf_fsaverage5(data_dir=None, url=None, resume=True, verbose=1):
+    """ Download Freesurfer fsaverage5 surface
+    Parameters
```
circle ci error can be fixed by giving an empty space between one line description and Parameters.

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+                 phenotypic=phenotypic,
+                 description=fdescr)
+
+def fetch_surf_fsaverage5(data_dir=None, url=None, resume=True, verbose=1):
```
PEP8. two empty lines required between functions.

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
                  description=fdescr)
+
+
+def fetch_surf_nki_enhanced(n_subjects=30, data_dir=None,
```
May be we need reduce default size in ```n_subjects``` to reduce the burden on CircleCI build ? 10 would be good I guess.

KamalakerDadi commented on doc/modules/reference.rst:
```diff
@@ -84,6 +84,9 @@ uses.
    fetch_nyu_rest
+   fetch_surf_nki_enhanced
+   fetch_surf_fsaverage5
+   fetch_atlas_surf_destrieux
```
You may have missed adding all "plotting" tools here. Like what you have done with fetchers.

KamalakerDadi commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,58 @@
+
+See :ref:`plotting` for more details.
+
+References
```
For me citing Destrieux paper would be enough I guess. Since other papers are already included in description in ```destrieux_surface.rst```

KamalakerDadi commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,58 @@
+"""
+Demonstrate basic loading and plotting of a cortical surface atlas.
```
Full stop (```.```) is not necessary in the title/headline. I haven't seen anything of that sort with other examples.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    # Download annot files, fsaverage surfaces and sulcal information
+    annot_file = '%s.aparc.a2009s.annot'
+    annot_url = url + '%i/%s.aparc.a2009s.annot'
+    annot_nids = {'lh annot': 9343, 'rh annot':9342}
```
PEP 8, missing white space here 'rh annot':9342 -> 'rh annot': 9342

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,63 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    for hemi in [('lh', 'left'), ('rh', 'right')]:
+
+        annot = _fetch_files(data_dir,
+                            [(annot_file % (hemi[1]),
```
Missing indent at [(annot_file)......]

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+    -------
+    data: sklearn.datasets.base.Bunch
+        Dictionary-like object, the interest attributes are :
+         - 'func': Paths to functional resting-state images
```
From what I see in description file and also return line. You are returning ```func_left``` and ```func_right```. This needs to be updated I guess.

KamalakerDadi replied:
Also, worth to write that the images we return are Gifti-like images.

I would write
```python 
- 'func_right' : Gifti-like images
      Paths to functional resting-state images
```

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+    dataset_name = 'nki_enhanced_surface'
+    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir,
+                                verbose=verbose)
+    ids = ['A00028185', 'A00033747', 'A00035072', 'A00035827', 'A00035840',
```
I am trying to download files. Id: 'A00028185', 'A00033747', seems to download well.

But, I have error fetching 'A00035072'. May this subject id is not available ?

```python
URLError: <urlopen error EOF occurred in violation of protocol (_ssl.c:590)| Error while fetching file A00035072_rh_preprocessed_fsaverage5_fwhm6.gii; dataset fetching aborted.>
```

KamalakerDadi replied:
Ok, now its seems downloading well 'A00035072' both 'lh' and 'rh'. Sorry about previous error comment.

juhuntenburg commented:
I have addressed all of @KamalakerDadi 's recent comments, apart from loading the .annot files, instead of returning their paths, in nilearn/datasets/atlas.py, as I wasn't sure if it is OK to import nibabel to do so.

juhuntenburg commented:
Am I correct that the main things to do now are deciding how to define the surface object and writing tests?

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,112 @@
+plotting.plot_surf_stat_map(fsaverage5_pial, stat_map=stat_map, hemi='left',
+                            bg_map=sulcal_depth_map, bg_on_data=True,
+                            cmap='Spectral', threshold=.6, alpha=.5,
+                            output_file='/tmp/plot_surf_stat_map.png')
```
The example should write in the current directory, and not in /tmp: /tmp might not exist (eg on Windows).

GaelVaroquaux commented:
> Am I correct that the main things to do now are deciding how to define the surface object and writing tests?

I believe that you are correct. Hurray!

With regards to the surface object, what's your proposition? I'd
like to keep it as simple as possible. Something like a list of files.

However, I looked at the examples of your PR, and I am not sure that we
want to do more than what's already in there. The examples are well
readable for me, and that's usually my criterion for API design.

Any further API choices should probably be done iteratively, based on
using this code.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,112 @@
+
+###############################################################################
+# Load resting state time series and parcellation
+timeseries = plotting.surf_plotting.check_surf_data(resting_state)
```
If we end up using "check_surf_data" in the examples, we should probably rename it "load_surf_data", so that it makes sense to the user.

KamalakerDadi replied:
+1

GaelVaroquaux commented on nilearn/datasets/func.py:
```diff
@@ -1669,3 +1669,219 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+
+    url: string, optional
+        Override download URL. Used for test only (or if you setup a mirror of
+        the data). Default: None
```
This comment should probably be addressed.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,68 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    ----------
+    data_dir: string, optional
+        Path of the data directory. Use to force data storage in a non-
+        standard location. Default: None
```
Empty line missing below.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,68 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+        Download URL of the dataset. Overwrite the default URL.
+    
+    resume: bool, optional (default True)
+        If true, try resuming download if possible.
```
I would prefer true -> True

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,68 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+
+    Parameters
+    ----------
+    data_dir: string, optional
```
Can we keep string -> str ?

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -782,3 +782,68 @@ def fetch_coords_dosenbach_2010(ordered_regions=True):
+    data_dir: string, optional
+        Path of the data directory. Use to force data storage in a non-
+        standard location. Default: None
+    url: string, optional
```
Same here string -> str

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1670,3 +1670,230 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+        By default, 10 subjects will be loaded. If None is given,
+        all 102 subjects will be loaded.
+
+    data_dir: string, optional
```
string -> str

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1670,3 +1670,230 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+        Path of the data directory. Used to force data storage in a specified
+        location. Default: None
+
+    url: string, optional
```
string -> str

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1670,3 +1670,230 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+    """ Download Freesurfer fsaverage5 surface
+    Parameters
+    ----------
+    data_dir: string, optional
```
string -> str

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1670,3 +1670,230 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+        Path of the data directory. Used to force data storage in a specified
+        location. Default: None
+
+    url: string, optional
```
same here, string -> str

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1670,3 +1670,230 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
+    -------
+    data: sklearn.datasets.base.Bunch
+        Dictionary-like object, the interest attributes are :
+         - 'pial_left': pial surface mesh for the left hemisphere
```
May we should express that files returned are gifti files ?

KamalakerDadi commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,49 @@
+# Display Destrieux parcellation on fsaverage5 surface
+plotting.plot_surf_roi(fsaverage5_pial, roi_map=parcellation[0],
+                       hemi='left', view='lateral', bg_map=sulcal_depth_map,
+                       bg_on_data=True, darkness=.5, cmap='gist_ncar')
```
May be adding ```title``` to the plots ?

GaelVaroquaux commented:
> May we should express that files returned are gifti files ?

+1

GaelVaroquaux commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+Only matplotlib is required.
+"""
+
+from nilearn._utils.compat import _basestring
```
We should probably use relative imports here: "from .._utils.compat import _basestring"

Also, these imports should be moved to the end of the list of imports, as imports should be ordered from general (standard library) to local.

GaelVaroquaux commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+                       alpha='auto', vmax=None, cmap='coolwarm',
+                       symmetric_cbar="auto", bg_on_data=False, darkness=1,
+                       output_file=None, **kwargs):
+
```
This empty line should be removed.

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+            surf_map_data = check_surf_data(surf_map)
+            if len(surf_map_data.shape) is not 1:
+                raise ValueError('surf_map can only have one dimension but has'
+                                 '%i dimensions' % len(surf_data_data.shape))
```
Seems like it is ```surf_map_data``` ?

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+
+
+# function to figure out datatype and load data
+def check_surf_data(surf_data):
```
Could you please add documentation ?

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+                surf_mesh.endswith('inflated')):
+            coords, faces = nibabel.freesurfer.io.read_geometry(surf_mesh)
+        elif surf_mesh.endswith('gii'):
+            coords, faces = gifti.read(surf_mesh).getArraysFromIntent(nibabel.nifti1.intent_codes['NIFTI_INTENT_POINTSET'])[0].data, \
```
[PEP 8] -> line too long

juhuntenburg replied:
I'm not sure how to break these lines correctly

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+                data[:, arr] = gii.darrays[arr].data
+            data = np.squeeze(data)
+        else:
+            raise ValueError('Format of data file not recognized.')
```
I would write more informative error message. "Format of data file is not recognized. Expected file formats are .nii, .nii.gz, ........ You provided .format(surf_data)"

juhuntenburg replied:
Good point, but how would I find out the format that is given if it's none of the ones we support? Just print out the ending of the filename?

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+            coords, faces = gifti.read(surf_mesh).getArraysFromIntent(nibabel.nifti1.intent_codes['NIFTI_INTENT_POINTSET'])[0].data, \
+                            gifti.read(surf_mesh).getArraysFromIntent(nibabel.nifti1.intent_codes['NIFTI_INTENT_TRIANGLE'])[0].data
+        else:
+            raise ValueError('Format of mesh file not recognized.')
```
Same comment as above here. More informative error message.

KamalakerDadi commented on nilearn/plotting/surf_plotting.py:
```diff
@@ -0,0 +1,376 @@
+            raise ValueError('Format of data file not recognized.')
+    # if the input is a numpy array
+    elif isinstance(surf_data, np.ndarray):
+        data = np.squeeze(surf_data)
```
Can we raise an error directly here rather than inside ?

```python
else:
      raise ValueError(Format of the given data file is not recognized. Expected file formats are .nii, .nii.gz, ........ You provided .format(surf_data))
```

From what I understood, only two types are accepted, either string or numpy array.

juhuntenburg replied:
But then I guess change the erro message? Because "file format not understood" doesn't really make sense when the input was a numpy array?

GaelVaroquaux commented:
> +            coords, faces = gifti.read(surf_mesh).getArraysFromIntent(nibabel.nifti1.intent_codes['NIFTI_INTENT_POINTSET'])[0].data, \

> I'm not sure how to break these lines correctly

I would do after the opening parenthesis of ".getArraysFromIntent("

GaelVaroquaux commented:
> Good point, but how would I find out the format that is given if it's none of
> the ones we support? Just print out the ending of the filename?

The full filename. It adds value to the user trying to debug (think a
large script loading many files at once).

GaelVaroquaux commented:
> But then I guess change the erro message? Because "file format not understood"
> doesn't really make sense when the input was a numpy array?

Yes, you can say: "the input type is not recognized. %r was given, while
valid inputs are..." and you list inputs that are supported.

juhuntenburg commented:
Regarding surface object: A surface mesh is now a list of two numpy arrays, the first contains the x,y,z coordinates of the mesh vertices, and the second contains the mesh faces (as indices into the first array). Seems like a simple and clean solution for the time being. I also allowed for inputting a mesh directly as such a list, in addition to providing a file. In that way one can even use the functions if working with surface formats that are not (yet) supported by the loading function (like myself). 

Regarding loading the annot file in the plot_surf_atlas example: When loading the Freesurfer annot file with Nibabel, one gets a list containing two arrays and a list. The first array contains for each vertex of the surface a value which is an index into the third list, that contains the regions labels as strings. But then there is the second array, which is a quirky Freesurfer thing, containing for each label RGBA values, according to Freesurfer conventions, and a number that I cannot find out the meaning of. Should I just drop the second array and return the index arrays as "map" and the label list as "labels" as done for aal?

juhuntenburg commented:
Also, is there a way for me to understand why the Travis CI build failed?

juhuntenburg commented:
I also moved the fsaverage5 dataset to struct.py, because it isn't actually functional at all. Does that make sense?

GaelVaroquaux commented:
> Also, is there a way for me to understand why the Travis CI build failed?

Memory profiler is now Python 2.6 incompatible.

We need to drop Python 2.6 support.

You can discard this failure.

GaelVaroquaux commented:
> I also moved the fsaverage5 dataset to struct.py, because it isn't actually functional at all. Does that make sense?

+1. Yes!

juhuntenburg commented:
Also regarding annot: it will still be two maps and two label lists for the two hemispheres, so either map_left, map_right, labels_left, labels_right, or a list [map, labels] for each hemisphere?

GaelVaroquaux commented:
> Regarding surface object: ... 

Sounds good to me.
 
> Regarding loading the annot file [...] Should I just drop the second array and return the index arrays as "map" and the label list as "labels" as done for aal?

I think so.

GaelVaroquaux commented:
> Also regarding annot: it will still be two maps and two label lists for the two
> hemispheres, so either map_left, map_right, labels_left, labels_right, or a
> list [map, labels] for each hemisphere?

What would be the benefit of having two maps and two lists? I don't use
surface, so I don't know.

juhuntenburg commented:
The two hemispheres are usually treated separately, since they are not a continuous mesh, so I think it would be rather confusing to combine them. If you look at the fsaverage5 and nki_enhanced dataset, the left and right hemisphere are always seperate

juhuntenburg commented:
Actually, i realize the label lists would be the same, so i guess we could do "labels", "map_right", "map_left"?

GaelVaroquaux commented:
OK, I had a look at the examples, and given that currently the plot function plot separately one hemisphere or the other, I think that we should go with separate lists.

To me, it's quite unsatisfying, as we are exposing details that are way too low level. I should be able to manipulate a "brain", and pass it to a plotting device that plots it, and choose if I want left or right hemisphere or both.

But that's for future work. We need to merge this and move forward before rethinking the world :).

juhuntenburg commented:
Alas, no rescuing the messy surface world yet.

juhuntenburg commented:
Ok, as for tests, that would be testing the load_surf_data and load_surf_mesh functions, to see they handle different input appropriately? The plot_surf function I don't really know how to test

juhuntenburg commented:
Ok, as for tests, that would be testing the load_surf_data and load_surf_mesh functions, to see they handle different input appropriately? The plot_surf function I don't really know how to test

GaelVaroquaux commented:
> Ok, as for tests, that would be testing the load_surf_data and load_surf_mesh
> functions, to see they handle different input appropriately?

Yes

> The plot_surf function I don't really know how to test

Have a look at how we test the plot_map functions. It's just smoke tests:
we don't check the results, but we do exercise the code paths.

KamalakerDadi commented:
plotting functions would be only smoke tests unless there are error messages which needs to tested if they are raised or not. For example using ```assert_raises_regex```.

juhuntenburg commented:
I have started to write tests, for now only for the loading functions. Still lacking testing of errors for load_surf_mesh, but if someone could have a look if that is going in a sensible direction that would be great.
I didnt find an easy way to create .annot and .label files, therefore I included example data for those tests. They are small. Does that work?

juhuntenburg commented:
Hm, I cannot really figure out why the tests fail here, it says the test.label and test.annot files don't exist. I tried to follow the same pattern as in test_image.py and it works fine running the test with nosetests on my computer.

juhuntenburg commented:
Hm, I cannot really figure out why the tests fail here, it says the test.label and test.annot files don't exist. I tried to follow the same pattern as in test_image.py and it works fine running the test with nosetests on my computer.

GaelVaroquaux commented:
I didn't have time to look, but it's probably related to package data:
http://stackoverflow.com/questions/13307408/python-packaging-data-files-are-put-properly-in-tar-gz-file-but-are-not-install
http://python-packaging.readthedocs.io/en/latest/non-code-files.html

juhuntenburg commented:
Thanks, found where to include it in the setup file.

juhuntenburg commented:
Yay! Will write smoke tests for the plotting functions, I think that's the only thing that's missing (?)

juhuntenburg commented:
Ah, I guess I also need to write tests for the data fetchers?

GaelVaroquaux commented on nilearn/datasets/tests/test_atlas.py:
```diff
@@ -417,3 +417,11 @@ def test_fetch_atlas_allen_2011():
 
     assert_not_equal(bunch.description, '')
+
+
```
This needs to run without internet connection, hence you should use (as in the other tests of this file), the test set up / tear down:

<pre>
 @with_setup(setup_mock, teardown_mock)
 @with_setup(tst.setup_tmpdata, tst.teardown_tmpdata)
</pre>

juhuntenburg commented:
I want to add the test for the nki enhanced fetcher, too. But when I run datasets/test/test_func.py with nosetests (before making changes to it) I get an error for the cobra dataset fetcher:

----
ERROR: nilearn.datasets.tests.test_func.test_fetch_cobre

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/julia/workspace/nilearn/nilearn/datasets/tests/test_func.py", line 518, in test_fetch_cobre
    url=local_url)
  File "/home/julia/workspace/nilearn/nilearn/datasets/func.py", line 1649, in fetch_cobre
    sz_ids = csv_array[csv_array['sz'] == 1.]['id'][:n_sz]
TypeError: slice indices must be integers or None or have an __index__ method
-------------------- >> begin captured stdout << ---------------------
Downloading data from whatever ...

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 13 tests in 0.424s

FAILED (errors=1)

juhuntenburg commented:
Hm, I am not sure I understand the the test set up / tear down and why the destrieux test fails now. Can someone give me a hint what I am missing?

GaelVaroquaux commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,49 @@
+###############################################################################
+from nilearn import plotting
+from nilearn import datasets
+import nibabel as nb
```
I don't think that we still need this import in the example.

GaelVaroquaux commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,49 @@
+plotting.plot_surf_roi(fsaverage5_pial, roi_map=parcellation,
+                       hemi='left', view='lateral', bg_map=sulcal_depth_map,
+                       bg_on_data=True, darkness=.5, cmap='gist_ncar')
+
```
Nitpick: I think that it would be better to add a split above this to have only one plot per cell.

GaelVaroquaux commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,49 @@
+# Fsaverage5 left hemisphere surface mesh files
+fsaverage5_pial = fsaverage['pial_left']
+fsaverage5_inflated = fsaverage['infl_left']
+sulcal_depth_map = fsaverage['sulc_left']
```
We have found that it's useful to print the value of these variables so that people understand what they are (filepaths). Indeed, people don't realize that they just need to give file names to the functions.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+"""
+Demonstrate basic loading and plotting of cortical surface data
+===============================================================
```
Missing empty line here.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+"""
+Demonstrate basic loading and plotting of cortical surface data
```
I think that the title of this example could be improved: the example isn't that basic: it included array manipulation to build a seed-based regression.

Maybe I would suggest "Seed-based connectivity on the surface"

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+The :func:`nilearn.plotting.plot_surf_stat_map` function is used
+to plot the resulting statistical map on the (inflated) pial surface.
+
+See :ref:`plotting` for more details.
```
It would be useful to add a see also pointing to the other seed-based connectivity http://nilearn.github.io/auto_examples/03_connectivity/plot_seed_to_voxel_correlation.html and add a see also on the other example pointing here.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+import numpy as np
+
+###############################################################################
+# Retrieve the data
```
It would be nice to add titles, basically underline the above, and similar on each cell. It makes the example easier to read.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+# Destrieux parcellation left hemisphere in fsaverage5 space
+destrieux_atlas = datasets.fetch_atlas_surf_destrieux()
+parcellation = destrieux_atlas['map_left']
+labels = destrieux_atlas['labels']
```
Same remark on printing a few of these variables and adding a comment pointing out that these are just file names.

GaelVaroquaux commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,122 @@
+stat_map[np.where(np.mean(timeseries, axis=1) == 0)] = 0
+
+###############################################################################
+# Display ROI on surface
```
It might be useful to split the cell below is several smaller cells: it's easier to comprehend.

GaelVaroquaux commented on nilearn/plotting/tests/data/test.label:
```diff
@@ -0,0 +1,328 @@
+#!ascii label  , from subject fsaverage5 vox2ras=TkReg 
```
Do we need that full file, or can we have a shortened version of it?

GaelVaroquaux commented on nilearn/plotting/tests/test_surf_plotting.py:
```diff
@@ -0,0 +1,313 @@
+def test_plot_surf_roi_error():
+    mesh = _generate_surf()
+    roi1 = np.random.randint(0, mesh[0].shape[0], size=5)
+    roi2 = np.random.randint(0, mesh[0].shape[0], size=10)
```
Here you are using the global, unseeded PRNG. We try to always avoid this, and to use a local PRNG created by rng = np.random.RandomState(0)

GaelVaroquaux commented:
I've made a bunch of comments, but they are super minor.

The only major issue that I see is that the tests are failing on travis and appveyor

I see them failing for several reasons:
* https://travis-ci.org/nilearn/nilearn/jobs/197929620 seems like a fairly trivial reason that should be easy to fix
* https://travis-ci.org/nilearn/nilearn/jobs/197929617 is due to a matplotlib version issue. We should investigate what maptlotlib version is needed and consider raising the dependency. Could you please find out what is the oldest matplotlib version that the code can work with
* https://travis-ci.org/nilearn/nilearn/jobs/197929614 seems to be due to a nibabel version. Same thing here: it would be useful to find which is the oldest nibabel that we can work with and consider whether or not we want to bump the requirement.

juhuntenburg commented:
Thanks! I have tried to address the small comments (not entirely sure if I got the "see other example" part right though). 

Regarding https://travis-ci.org/nilearn/nilearn/jobs/197929620 : this happens since I set the test set up / tear down (see my last comment), the annot data seems not to be found / loaded anymore. Does that have something to do with the annotation file actually being loaded into memory by this datafetcher? I works fine if i remove the decorators

` @with_setup(setup_mock, teardown_mock)
 @with_setup(tst.setup_tmpdata, tst.teardown_tmpdata)`

For the dependencies: what is the most efficient way to test which are the oldest versions that work?

KamalakerDadi commented:
Two failures with respect to version dependency nibabel 1.1.0. ```freesurfer``` module is not implemented in nibabel 1.1.0 and also class ```gifti.GiftiDataArray``` has no argument ```intent```. Hence tests failed with nibabel version dependency 1.1.0.

I am not sure of another failure ```Key Error: <type 'int'>``` when writing empty dictionary to gifti file. Can you please have a look ?

May be we could do something like this to fix the tests:
```python
if LooseVersion(nibabel.__version__) == LooseVersion('1.1.0'):
    # Old nibabel cannot use freesurfer module to load the surf atlases
    # Therefore, we
    raise SkipTest
```
We need to add as a "Note" in documentation of atlas fetchers as well as in plotting examples about minimum version requirement of nibabel. This could be as a warning for users particularly who use nibabel==1.1.0. Do you agree ?

I will keep updating with another version failures if any ?

GaelVaroquaux commented:
> Two failures with respect to version dependency nibabel 1.1.0. freesurfer
> module is not implemented in nibabel 1.1.0 and also class gifti.GiftiDataArray
> has no argument intent. Hence tests failed with nibabel version dependency
> 1.1.0.

What's the minimum version of nibabel required? Could one of you do a
bisection on the versions of nibabel to find this out?

KamalakerDadi commented:
>What's the minimum version of nibabel required? Could one of you do a bisection on the versions of nibabel to find this out?

1.2.0
Ref: http://nipy.org/nibabel/changelog.html#changelog

GaelVaroquaux commented:
> 1.2.0
> Ref: http://nipy.org/nibabel/changelog.html#changelog

OK, we should bump the requirement to nibabel 1.2.0.

Kamalakar, do you want to do it (change the CI and the specified versions
in the docs and in our checks, grep the code to make sure that we don't
have compatibility layers that are lying around for version 1.1.0), and send
a PR. We'll review it, merge it, and Julia can then rebase on master to
get rid of these issues.

KamalakerDadi commented:
>Kamalakar, do you want to do it (change the CI and the specified versions in the docs and in our checks, grep the code to make sure that we don't have compatibility layers that are lying around for version 1.1.0), and send a PR. We'll review it, merge it, and Julia can then rebase on master to get rid of these issues.

I am going to try it. Soon do a PR.

juhuntenburg commented:
Hi, thanks, fantastic! Can anyone give me a hint on what might be going on with the destrieux datafetcher test, i.e. what could be the reason it fails since I added the test setup / teardown decorators? 

Also, @KamalakerDadi did you get my e-mail on the report?

KamalakerDadi commented:
Bumping the requirement to 1.2.0 can fix freesurfer but not argument failure ```intent``` from class gifti.GiftiDataArray ? Any hints with @juhuntenburg

KamalakerDadi commented:
>Also, @KamalakerDadi did you get my e-mail on the report?

I got it. Thanks. I need to go through it.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,72 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+                      left hemishpere of the fsaverage5 surface
+        - 'map_right': Numpy array, index into 'labels' for each vertex on the
+                       right hemishpere of the fsaverage5 surface
+
```
Documentation for 'description' is missing here.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,72 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+    -------
+        data: sklearn.datasets.base.Bunch
+        dictionary-like object, contains:
+        - 'labels' : List of region labels
```
I am in favour of like this:

- 'labels' : list
            Contains labels defined to regions.

Same with 'map_left' : numpy.ndarray
                               index into 'labels' for each vertex on the left hemishpere
                               of the fsaverage5 surface

Also, could you correct spelling of "hemishpere" to "hemisphere"

KamalakerDadi replied:
Indents are missing in my previous comment. Also a worth to check. Useful for doc build with sphinx.

KamalakerDadi commented:
>Can anyone give me a hint on what might be going on with the destrieux datafetcher test, i.e. what could be the reason it fails since I added the test setup / teardown decorators?

I cannot reproduce the failure. May be I am missing something.

I think it is better to rebase after NiBabel bump PR is merged to get rid of atleast some failures and then fix remaining ones.

juhuntenburg commented:
... working on the intent problem

KamalakerDadi commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,57 @@
+      fsaverage['sulc_left'])
+
+###############################################################################
+# Display Destrieux parcellation on fsaverage5 surface
```
Saying that it's "pial" is missing in the comment. Like as you said below with inflated.

juhuntenburg replied:
Done

KamalakerDadi commented on examples/01_plotting/plot_surf_atlas.py:
```diff
@@ -0,0 +1,57 @@
+# The parcellation is already loaded into memory
+parcellation = destrieux_atlas['map_left']
+
+# Retrieve fsaverage surface template
```
What you retrieve here is bunch of templates. May be we need to update that we retrieve templates and then pick one specific template as a background for plotting. For clarity sake.

juhuntenburg replied:
Not sure if I understand. It is the same template, just once inflated and once not inflated. Should I add that?

KamalakerDadi replied:
Yeah I agree. But you have bunch of templates in fsaverage not one template. I would just write "Retrieving fsaverage5 surface based templates and use them for plotting"

juhuntenburg replied:
Done

juhuntenburg replied:
Oh didnt see your comment. I did something slightly different: 
"Retrieve fsaverage5 surface dataset, which contains pial and inflated version of the surface template as well as sulcal depth maps"

juhuntenburg commented:
Regarding the `intent` argument failure -- writing data with a specific intent to a gifti file seems to only work in the newest Nibabel release (2.1). I only use it in this test:  https://github.com/nilearn/nilearn/pull/1016/files#diff-c96717360f79551c86a356429012a22fR133
So two options:
1. I create the test gifti test files and add them to the test data, instead of creating them within the test
2. Don't test the loading functions to recognize missing intents when loading gifti

KamalakerDadi commented:
>Don't test the loading functions to recognize missing intents when loading gifti

If I understood correctly, you want to test the function only for newest version. i.e. nibabel 2.1.
Using ```LooseVersion``` can be helpful if I am correct.
```python
from distutils.version import LooseVersion
```

juhuntenburg commented:
No, only the test requires Nibabel 2.1, not the actual function, so I figure it would be better to change the test itself to not be dependent on Nibabel 2.1. Could be achieved by the above options. I'd vote for just storing a small test gifti

KamalakerDadi commented:
Same as like with argument ```intent```, we have similar no attribute errors with ```write_morph_data``` and ```write_geometry```with older versions.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,77 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+        standard location. Default: None
+
+    url: str, optional
+        Download URL of the dataset. Overwrite the default URL.
```
Empty line missing below.

juhuntenburg commented:
Same solutions apply, I only use them in the tests to write intermediate files to test the loading function. Can prevent that by adding small test files. Should I?
Or rather only run those tests if Nibabel 2.1 is installed?

KamalakerDadi commented:
I would rather say doing version check and run tests passing with latest nibabel versions. As users now a days are mostly updating to most recent versions. Mostly, few users will have older versions. We can also have opinions from @GaelVaroquaux @AlexandreAbraham

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,77 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+
+    Returns
+    -------
+        data: sklearn.datasets.base.Bunch
```
Here, we have some documentation misalignent problems.

 "data" can be aligned right below ----------- and missing "empty line" between (dictionary-like object, contains:) and (- 'labels' : list) and no space before colon for return argument 'labels' and missing empty line between each return argument.

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,77 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+                               resume=True, verbose=1):
+    """Download and load Destrieux et al, 2010 cortical atlas.
+
+    This atlas returns 74 sulco-gyral structures per hemisphere distributed
```
Actually, I saw tests you test with 76 labels but you have written 74 here ? Can you tell me difference ?

juhuntenburg replied:
One label is "Unknown" and one is "Medial wall", so only 74 real labels. Still seems kind of confusing, I'll fix it

KamalakerDadi commented on nilearn/datasets/atlas.py:
```diff
@@ -863,3 +863,77 @@ def fetch_atlas_allen_2011(data_dir=None, url=None, resume=True, verbose=1):
+    -------
+        data: sklearn.datasets.base.Bunch
+        dictionary-like object, contains:
+        - 'labels' : list
```
I just saw other fetchers. They have return arguments in double quotes but not single quotes. May be that's why sphix fails. Can you also correct them ?

KamalakerDadi commented on nilearn/datasets/tests/test_atlas.py:
```diff
@@ -417,3 +417,13 @@ def test_fetch_atlas_allen_2011():
+def test_fetch_atlas_surf_destrieux(data_dir=tst.tmpdir, verbose=0):
+    bunch = atlas.fetch_atlas_surf_destrieux()
+    assert_equal(len(bunch.labels), 76)
+    assert_equal(bunch.map_left.shape, (10242,))
```
Could you tell me why index to labels are 10242 but not 76 ?

KamalakerDadi commented:
Reminder: Just in case, if you have forgot to add tests for NKI fetcher.

GaelVaroquaux commented:
I just did a PR to this branch to fix the test problem: https://github.com/juhuntenburg/nilearn/pull/2

We also need similar tests for NKI. @KamalakerDadi is going to try to have a go at them.

GaelVaroquaux commented:
Also, we still have some test failing probably due to different versions of the dependencies. We need to fix them:
https://travis-ci.org/nilearn/nilearn/builds/203910386

That's why tests are useful :)

juhuntenburg commented:
@GaelVaroquaux  Yes, I believe this is all Nibabel and all related to creating the test data during the tests for the loading function (gifti and freesurfer type). @KamalakerDadi suggested to skip the tests if Nibabel is < 2.0. Alternatively I could create the test data beforehand and put it in ../plotting/tests/data/ .

What do you prefer?

juhuntenburg commented:
@KamalakerDadi I just addressed your small comments regarding sphinx, let me know if that fixes it

KamalakerDadi commented:
I am going to add tests for NKI fetcher.

juhuntenburg commented:
Ah sorry, just saw your comment @KamalakerDadi !
I added a test, but not entirely sure if this is how it should be. Maybe you can have a look and take it from there?

GaelVaroquaux commented:
So, upon inspection of the test failures, the code cannot work with matplotlib 0.11, but we still need to support matplotlib 0.11 because it's the install at neurospin (don't get me started on that).

So, I suggest that we support surface plotting only if matplotlib is recent enough. It means testing for the version of matplotlib in plotting/__init__ and in the tests to skip some tests. This will be required to get travis to pass.

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,149 @@
+timeseries = plotting.surf_plotting.load_surf_data(nki_dataset['func_left'][0])
+
+# Extract seed region via label
+pcc_region = 'G_cingul-Post-dorsal'
```
Can you add ```b'G_cingul-Post-dorsal'``` to make it to unicode. If not fails with Python 3.

juhuntenburg replied:
done

KamalakerDadi commented on examples/01_plotting/plot_surf_stat_map.py:
```diff
@@ -0,0 +1,149 @@
+# --------------------------------
+
+# Load resting state time series
+timeseries = plotting.surf_plotting.load_surf_data(nki_dataset['func_left'][0])
```
```load_surf_data``` data loader should be in datasets.py or no ?

juhuntenburg replied:
no i dont think so

KamalakerDadi commented:
>I added a test, but not entirely sure if this is how it should be. Maybe you can have a look and take it from there?

Tests you have added are good. I think its no need to take.

KamalakerDadi commented:
Can you please rebase onto master ?

GaelVaroquaux commented:
> Tests you have added are good. I think its no need to take.

OK, so the remaining thing to do is to skip import and test of surface
plotting if matplotlib is not recent enough, and after we can merge.

KamalakerDadi commented:
>@KamalakerDadi I just addressed your small comments regarding sphinx, let me know if that fixes it

Sphinx issues are mostly of documentation. We can fix travis and then move on to sphinx.

juhuntenburg commented:
> @GaelVaroquaux Yes, I believe this is all Nibabel and all related to creating the test data during the tests for the loading function (gifti and freesurfer type). @KamalakerDadi suggested to skip the tests if Nibabel is < 2.0. Alternatively I could create the test data beforehand and put it in ../plotting/tests/data/ .

Still need to fix this, which solution do you prefer?

GaelVaroquaux commented:
I have worked around one nibabel problem in the tests:
https://github.com/juhuntenburg/nilearn/pull/3

@KamalakerDadi will take over and finish the details today, as we want to release today.

KamalakerDadi commented:
We decided to skip tests.

KamalakerDadi commented on nilearn/plotting/tests/test_surf_plotting.py:
```diff
@@ -0,0 +1,324 @@
+import numpy as np
```
imports needs to be reordered. More general to local.
import os
import tempfile

import nibabel
......

juhuntenburg commented:
Ok, i just rebased and fixed the unicode issue in the plot_surf_stat_map example. I have two other deadlines tomorrow, so if you want to release today you might need to take it from here

KamalakerDadi commented:
Thanks for fixing conflicts and rebasing. I will do the rest.

juhuntenburg commented:
Awesome, thanks!

GaelVaroquaux commented:
Closing this PR, as the corresponding functionality has been merged via #1387

Hurray; hurray, :beers:

juhuntenburg commented:
Yayyy!! Almost exactly one year :)
>  juhuntenburg commented on 25 Feb 2016
